exchange/index.go

```diff
@@ -453,6 +469,97 @@ func (idx *Index) evict(size uint64) uint64 {
 	return evicted
 }
 
+// tagForGC tags CIDs that will be evicted during garbage collection
+func (idx *Index) tagForGC(ref *DataRef) error {
+	idx.emu.Lock()
+	defer idx.emu.Unlock()
+
+	return utils.WalkDAG(context.TODO(), ref.PayloadCID, idx.bstore, sel.All(), func(block blocks.Block) error {
+		idx.gcSet.Add(block.Cid())
+		return nil
+	})
+}
+
+// GC removes tagged CIDs
+func (idx *Index) GC() error {
+	idx.emu.Lock()
+	defer idx.emu.Unlock()
+
+	// exit if there is nothing to evict
+	if idx.gcSet.Len() == 0 {
+		return nil
+	}
+
+	gcds, ok := idx.ds.(datastore.GCDatastore)
+	if !ok {
+		return errors.New("datastore is not a GCDatastore")
+	}
+	err := gcds.CollectGarbage()
+	if err != nil {
+		return err
+	}
+
+	gcbs, ok := idx.bstore.(blockstore.GCBlockstore)
+	if !ok {
+		return errors.New("blockstore is not a GCBlockstore")
+	}
+
+	unlock := gcbs.GCLock()
+	defer unlock.Unlock()
+
+	err = idx.gcSet.ForEach(func(c cid.Cid) error {
+		return idx.bstore.DeleteBlock(c)
+	})
+	if err != nil {
+		return fmt.Errorf("failed to run garbage collector: %v", err)
+	}
+
+	idx.gcSet = cid.NewSet()
+
+	return nil
+}
+
+// CleanBlockStore removes blocks from blockstore which CIDs are not in index
+func (idx *Index) CleanBlockStore(ctx context.Context) error {
+	idx.emu.Lock()
+	defer idx.emu.Unlock()
+
+	cidSet := cid.NewSet()
+
+	err := idx.root.ForEach(ctx, func(k string, val *cbg.Deferred) error {
+		ref := new(DataRef)
+		err := ref.UnmarshalCBOR(bytes.NewReader(val.Raw))
+		if err != nil {
+			return err
+		}
+
+		return utils.WalkDAG(ctx, ref.PayloadCID, idx.bstore, sel.All(), func(blk blocks.Block) error {
+			cidSet.Add(blk.Cid())
+			return nil
+		})
+	})
+	if err != nil {
+		return err
+	}
+
+	kc, err := idx.Bstore().AllKeysChan(ctx)
+	if err != nil {
+		return err
+	}
+	for k := range kc {
+		key := cid.NewCidV1(cid.DagCBOR, k.Hash())
+		if cidSet.Has(key) {
+			continue
+		}
+		err = idx.Bstore().DeleteBlock(k)
+		if err != nil {
+			return err
+		}
+	}
```

Comment:
Also GC the datastore here. Any time you're deleting blocks basically.

Reply:
You missed that one no?

Reply:
Ah damn, I've done it but reverted everything because of failing tests and I forgot to put it back